src/components/tl/cuda/alltoallv/alltoallv_ce.c

```diff
@@ -0,0 +1,402 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021-2022.  ALL RIGHTS RESERVED.
+ * Copyright (c) Meta Platforms, Inc. and affiliates. 2022.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "alltoallv.h"
+#include "components/ec/ucc_ec.h"
+#include "tl_cuda_cache.h"
+#include "utils/arch/cpu.h"
+#include "utils/arch/cuda_def.h"
+
+enum {
+    ALLTOALL_CE_STAGE_SYNC,  /*< Wait for free SYNC segment */
+    ALLTOALL_CE_STAGE_SETUP, /*< Wait for memhandle setup to finish */
+    ALLTOALL_CE_STAGE_POST_COPIES,
+    ALLTOALL_CE_STAGE_COPY, /*< Wait for all copies to finish */
+    ALLTOALL_CE_STAGE_BAR,  /*< Wait for other ranks to finish */
+};
+
+ucc_status_t ucc_tl_cuda_alltoallv_ce_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_cuda_task_t *task = ucc_derived_of(coll_task, ucc_tl_cuda_task_t);
+
+    tl_trace(UCC_TASK_LIB(task), "finalizing task %p", task);
+    ucc_ec_destroy_event((void *)task->alltoallv_ce.copy_done,
+                         UCC_EE_CUDA_STREAM);
+    ucc_tl_cuda_task_put(task);
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_cuda_alltoallv_setup_start(ucc_tl_cuda_task_t *task)
+{
+    ucc_tl_cuda_team_t *team = TASK_TEAM(task);
+    ucc_tl_cuda_sync_t *sync = TASK_SYNC(task, UCC_TL_TEAM_RANK(team));
+    ucc_status_t        status;
+    ucc_coll_args_t    *args = &TASK_ARGS(task);
+
+    // for A2av: copy counts and displ. for proxies (if any) to access
+    if (team->topo->num_proxies > 0 &&
+        UCC_COLL_TYPE_ALLTOALLV == args->coll_type) {
+        size_t elem_size = UCC_COLL_ARGS_COUNT64(args) ? 8 : 4;
+        memcpy(sync->src_cnts, args->src.info_v.counts,
```

Comment:
it would be easier if we save offsets and counts in bytes rather than in elements, it will solver sdt_size and rdt_size problem and also enable support for the case when source datatype is not the same as recv datatype